DFE-Digital/sts-plan-technology-for-your-school
# Feat: Implement new sql caching

## Overview

Addresses ticket [#216205](https://dfe-ssp.visualstudio.com/s190-schools-technology-services/_workitems/edit/216205)

Changes our approach to caching SQL queries to avoid excess connections - as per [ADR 38](https://github.com/DFE-Digital/sts-plan-technology-for-your-school/blob/2f6f36134ba10ebfafd22b73bfc458da817e4322/docs/adr/0038-sql-caching-mechanism.md)

**Azure function unit tests are failing - but these are about to change, so holding off on the azure function PR for those**

## Changes

### Major
- We no longer use EFCoreSecondLevelCacheInterceptor, and have replaced this with a simple MemoryCache that uses the query string to cache the results for 30 minutes 
- The CmsDbContext now uses the cached methods for all queries
 
### Non-functional changes 

Updated docs to mention new process in the code conventions (this seems like the right place)

## How to review the PR

- Check that the caching works, you should now see the sql calls disappear entirely when a content only query is repeated.
- Review whether the documentation is sufficient for caching going forward.
- Add any thoughts on the approach

This now uses a `QueryCache` singleton class with all the cache methods injected into the `CmsDbContext`. So all cms queries need to be of the form
```csharp
_db.ToListAsync(_db.Thing)
```
rather than
```csharp
_db.Thing.ToListAsync()
```
This ensures the caching is only used here, and it only required updating a few places to achieve

## Checklist
